### DISCUSSION TITLE
Issues with MWAA accessing airflow/connections/aws_default

### DISCUSSION THREAD

**Discussion by rankol** [apache/airflow#24088] [Q&A] [answered]
Hello,

I am using MWAA (AWS Managed Airflow) and when I attempt to access a secret from Secrets Manager I get the error below in this gist. Additional logs show that the secret I actually want to retrieve were able to be printed in the logs.

https://gist.github.com/rankol/609b4e8a3fb04bd40344afc727857b5b

In my DAG I have a Python function that uses AwsBaseHook to access SecretsManager from Boto3.

https://gist.github.com/rankol/e22ee9cac98d678422aae37b8cebb0e2

I see in my Connections drop down on MWAA where I have aws_default without anything defined in there. Essentially I would like to figure out a way in my DAG to not attempt accessing airflow/connections/aws_default since this causes alerts to be created. Is there a fallback credential that is used? I was able to print out my target secret.

If I use the SecretsManager hook does that help or would I come across a similar issue since that uses AwsBaseHook as its base also?

https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/secrets_manager/index.html#airflow.providers.amazon.aws.hooks.secrets_manager.SecretsManagerHook

The only way I have gotten this error to not be present is to set the execution role's IAM policy to `Resource:*` but that allows all secrets to be accessed and I am attempting to limit what secrets can be retrieved. 

If I set the IAM policy to something such as the two options in this block of code I still get the errors mentioned in the first link.

```
"Resource": "arn:aws:secretsmanager:us-east-1:<ACCOUNT>:secret:aws_default or airflow/connections/aws_default
```

**Comment by potiuk** [comment]
I think this shoudl be asked at MWAA support, there are far too many MWAA specific things in here, this is really an MWAA question, not airflow one.

**Reply by rankol** [reply]
Thanks, I'll reach out to AWS support and see what they say.

**Reply by ajf-firstup** [reply]
@rankol Did you ever open a support case? Did you get any further info? We're experiencing the same log flooding here. Thanks!

**Reply by Taragolis** [reply]
https://github.com/apache/airflow/discussions/24088#discussioncomment-2867619

**Reply by ajf-firstup** [reply]
Thanks! That resolves permission issues; in our case, we're trying to resolve "ResourceNotFoundException" -- the MWAA instance has permission to request the secret, but the secret (intentionally!) does not exist. (Our corp policy is to never have "default" connections -- whether this is too strict of an interpretation in this point is debatable, but...)

For one instance, I just stuck an empty JSON object into that secret, and it stopped the complaints -- we're just trying to understand if there are downstream implications to doing this.

Ideally there'd be a way to configure MWAA / Airflow to not try to make this connection at all, but @Taragolis's comments below make me think that's not likely:
> Due to current implementation of Amazon Provider Hooks, it always fallback to the default boto3 strategy if connection not exists (it would changed in the future and would raise actual error).

Re-reading that, I think we'd rather have the explicit error.

Thanks for the info, regardless!

**Reply by Taragolis** [reply]
> the MWAA instance has permission to request the secret, but the secret (intentionally!) does not exist.

If connection/variable not found in one of the source it try to another. The lookup order for connections/variables are:
1. Alternate Secret Backend, if it configured
2. Environment variables
3. Airflow Database

> we're trying to resolve "ResourceNotFoundException" -- the MWAA

As I could see the error should be printed only in `DEBUG` severity

https://github.com/apache/airflow/blob/83bd60fd97d4ca622adcbd7898d88880fee43054/airflow/providers/amazon/aws/secrets/secrets_manager.py#L315-L321

**Reply by ajf-firstup** [reply]
> The lookup order for connections/variables are:
>
> 1. Alternate Secret Backend, if it configured
> 2. Environment variables
> 3. Airflow Database

Thank you -- this summary is helpful, and I'll see what that team is actually using.

(In particular, if I created the secret and just put an empty JSON object as the value, would that cause any downstream problems? I'll try to trace down the code from the point you gave above!)

> As I could see the error should be printed only in `DEBUG` severity

The error from **Airflow** might only be at DEBUG.

But even *trying* to access the secret is an AWS secretsmanager:GetSecretValue call -- and all those are logged by CloudTrail. Normal accesses are fine, but when a non-existent secret is requested, it's an error (resource not found), and captured as a critical error against SecretsManager. With a few dozen DAGs, we're seeing O(100k) of these errors per day per MWAA instance.

(And inspecting accesses to non-existent secrets is a valid security exercise, since it could represent an attacker probe.)

I appreciate that we can work around this multiple ways, but given how intricate Airflow is, and how defaults interact, we're reaching out to to anyone that's run into similar problems. :-)

Thanks again for your replies!

**Comment by rankol** [comment, marked as answer]
I was able to resolve this by setting `"Resource": "arn:aws:secretsmanager:us-east-1:<ACCOUNT>:secret:airflow/connections/*"` in my execution role's IAM policy. Using `airflow/*/*` worked for me but I wanted to be a bit more restrictive if possible so I went with `airflow/connections/*`.

Setting `airflow/connections/aws_default` did not seem to work in my case.

I added the following permissions for access to that specific resource.

```
"secretsmanager:GetResourcePolicy",
"secretsmanager:GetSecretValue",
"secretsmanager:DescribeSecret",
"secretsmanager:ListSecretVersionIds"
```

**Reply by potiuk** [reply]
Cool. Thanks for that comment. Might be great for others looking for answers.

**Comment by manugarri** [comment]
adding up on this issue... What i dont understand, is why MWAA is trying to access `aws_default` when the instance is using IAM role.

Is there any way to setup a the Secretsmanager backend, but add an exception to the `aws_default` connection?

Getting the error on every dag log is a bit cumbersome.

**Reply by Taragolis** [reply]
What kind of error? Due to current implementation of Amazon Provider Hooks, it always fallback to the default boto3 strategy if connection not exists (it would changed in the future and would raise actual error).

But if you have particular problem with Amazon Provider you'd rather open new Discussion (if you not sure is this error or not) or Issue if you sure that this is a bug. Please provide in this case as much information as possible:  more information more chance that you would have answer for you question.

Please note if you think it is only related to AWS MWAA, you need to open Support Ticket in your AWS Account. Unlikely that community could help you with managed service.
